File: backend/Chord_detection/transpose.py
# Define the notes for transposition
NOTES = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"]
def get_semitone_input():
        """
        Prompts the user to input the number of semitones for transposition.
        """
        while True:
            try:
                semitones = 0 #int(input("Enter the number of semitones to transpose the chords (e.g., +2 or -2): "))
                return semitones
            except ValueError:
                print("Invalid input. Please enter an integer value.")

def transpose_chord(chord, semitones):
    """
    Transposes a single chord by the given number of semitones.
    """
    for i, char in enumerate(chord):
        if char.isdigit() or char in ["m", "M", "b", "#"]:
            continue
        root = chord[:i + 1]
        chord_type = chord[i + 1:]
        break
    else:
        root = chord
        chord_type = ""
    # Normalize double sharps (##) and double flats (bb)
    
    if "##" in root:
        root = NOTES[(NOTES.index(root.replace("##", "")) + 2) % len(NOTES)]
    elif "bb" in root:
        root = NOTES[(NOTES.index(root.replace("bb", "")) - 2) % len(NOTES)]
    
    # Handle enharmonic equivalents
    try:
        root_index = NOTES.index(root)
        transposed_index = (root_index + semitones) % len(NOTES)
        transposed_root = NOTES[transposed_index]
        return transposed_root + chord_type
    except ValueError:
        return chord  # Return the original chord if transposition fails

def transpose_chords(chords, semitones):
    """
    Transposes a list of chords by the given number of semitones.
    """
    transposed_chords = []
    for start_time, end_time, chord_label in chords:
        transposed_chord = transpose_chord(chord_label, semitones)
        transposed_chords.append((start_time, end_time, transposed_chord))
    return transposed_chords

File: frontend/src/utils/transpose.js
const NOTES = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];

export const transposeChord = (chord, semitones) => {
  if (!chord) return chord;
  let root = "";
  let chordType = "";

  // Parse the chord to separate root and chord type
  for (let i = 1; i <= chord.length; i++) {
    const potentialRoot = chord.substring(0, i);
    if (NOTES.includes(potentialRoot)) {
      root = potentialRoot;
      chordType = chord.substring(i);
      break;
    }
  }
  if (!root) {
    root = chord;
    chordType = "";
  }

  try {
    // Normalize double sharps (##) and double flats (bb)
    let normalizedRoot = root;
    while (normalizedRoot.includes("##")) {
      const baseNote = normalizedRoot.replace("##", "");
      const baseIndex = NOTES.indexOf(baseNote);
      if (baseIndex !== -1) {
        normalizedRoot = NOTES[(baseIndex + 2) % NOTES.length];
      } else {
        break;
      }
    }
    while (normalizedRoot.includes("bb")) {
      const baseNote = normalizedRoot.replace("bb", "");
      const baseIndex = NOTES.indexOf(baseNote);
      if (baseIndex !== -1) {
        normalizedRoot = NOTES[(baseIndex - 2 + NOTES.length) % NOTES.length];
      } else {
        break;
      }
    }

    // Transpose the normalized root
    const rootIndex = NOTES.indexOf(normalizedRoot);
    if (rootIndex === -1) return chord;

    let transposedIndex = (rootIndex + semitones + NOTES.length) % NOTES.length;
    let transposedRoot = NOTES[transposedIndex];

    // Additional normalization for the transposed result
    while (transposedRoot.includes("##")) {
      const baseNote = transposedRoot.replace("##", "");
      const baseIndex = NOTES.indexOf(baseNote);
      if (baseIndex !== -1) {
        transposedRoot = NOTES[(baseIndex + 2) % NOTES.length];
      } else {
        break;
      }
    }
    while (transposedRoot.includes("bb")) {
      const baseNote = transposedRoot.replace("bb", "");
      const baseIndex = NOTES.indexOf(baseNote);
      if (baseIndex !== -1) {
        transposedRoot = NOTES[(baseIndex - 2 + NOTES.length) % NOTES.length];
      } else {
        break;
      }
    }

    return transposedRoot + chordType;
  } catch (error) {
    console.error("Transposition error:", error);
    return chord;
  }
};

export const transposeChords = (chords, semitones) => {
  if (!chords || !Array.isArray(chords)) return [];

  return chords.map((chord) => ({
    ...chord,
    chord: transposeChord(chord.chord, semitones),
  }));
};